### ISSUE TITLE
Feature Request: RemoteAgentLoop - Support for External Distributed Agent Integration

### ISSUE THREAD

**Issue by KunWuLuan** [verl-project/verl#5737]
### Feature request

# Feature Request: RemoteAgentLoop - Support for External Distributed Agent Integration

## Summary

Request to add a new `RemoteAgentLoop` type that enables VeRL to integrate with external agents running in separate Ray or Kubernetes clusters. This allows RL training of third-party agents (e.g., SWE-agent, Claude Code, etc.) without requiring them to run within the VeRL infrastructure.

**Status**: Reference implementation testing in inner repo.

---

### Current Limitations

VeRL currently provides several AgentLoop implementations:

1. **SingleTurnAgentLoop** - Single-turn chat completion only
2. **ToolAgentLoop** - Multi-turn tool execution with local tools
3. **ReactAgentLoop** - LangGraph-based ReAct pattern
4. **CollabLLMAgentLoop** - Interaction with simulated users via LLM APIs

However, **none of these support**:

- Running agents in external/distributed environments (separate Ray/K8s clusters)
- Capturing token-level data (token_ids, logprobs) from black-box LLM calls
- Integrating production agents that manage their own LLM connections
- RL training of agents that cannot be modified to use VeRL's internal APIs

### Use Cases

1. **SWE-bench Training**: Run SWE-agent in isolated Docker environments while training on code repair tasks
2. **Multi-Agent Systems**: Coordinate agents across different orchestration frameworks
3. **Production RLHF**: Train on real user-facing agents without disrupting serving infrastructure
4. **Third-Party Agent Evaluation**: Benchmark and optimize external agent systems with RL

---

### Architecture Overview

<img width="1264" height="848" alt="Image" src="https://github.com/user-attachments/assets/53d2d64b-c9eb-4f54-9932-e41ef449b66e" />

```
┌─────────────────────────────────────────────────────────────────────────┐
│                         VeRL Training Cluster                            │
│                                                                          │
│  ┌──────────────────────┐                                               │
│  │  AgentLoop Worker    │                                               │
│  │  (Ray Actor)         │                                               │
│  │                      │                                               │
│  │  ┌────────────────┐  │   HTTP (OpenAI API)                          │
│  │  │ RemoteAgent    │◄─┼──────────────────────────────────────┐       │
│  │  │ Loop           │  │                                      │       │
│  │  └───────┬────────┘  │                                      │       │
│  │          │           │                                      │       │
│  │          ▼           │                                      │       │
│  │  ┌────────────────┐  │   ┌─────────────────────────────┐   │       │
│  │  │ LLMProxyServer │  │   │  Third-Party Agent Cluster  │   │       │
│  │  │ (Ray Singleton)│  │   │  (Ray or Kubernetes)        │   │       │
│  │  │                │  │   │                             │   │       │
│  │  │ - FastAPI      │  │   │  ┌──────────────────────┐  │   │       │
│  │  │ - Session Rec. │  │   │  │ Third-Party Agent    │  │   │       │
│  │  │ - Tokenizer    │  │   │  │ (e.g., SWE-agent)    │  │   │       │
│  │  └───────┬────────┘  │   │  │                      │  │   │       │
│  │          │           │   │  │ OpenAI SDK           │  │   │       │
│  │          │           │   │  │ base_url =           │  │   │       │
│  │          │           │   │  │ http://proxy/{id}/v1 │  │   │       │
│  │          │           │   │  └──────────┬───────────┘  │   │       │
│  │          │           │   │             │              │   │       │
│  │          ▼           │   └─────────────┼──────────────┘   │       │
│  │  ┌───────────────┐    │                 │                  │       │
│  │  │ vLLM Rollout  │    │                 │                  │       │
│  │  │ Servers       │    │                 │                  │       │
│  │  │ (server_      │    │                 │                  │       │
│  │  │   handles)    │    │                 │                  │       │
│  │  └───────────────┘    │                 │                  │       │
│  └──────────────────────┘                 │                  │       │
│                                           │                  │       │
│                                           │                  │       │
└───────────────────────────────────────────┼──────────────────┼───────┘
                                            │                  │
                                            │                  │
                                    ┌───────▼──────────────────▼───────┐
                                    │     Remote Infrastructure         │
                                    │                                   │
                                    │  Harbor Server / Custom Agent     │
                                    │  Runtime Environment              │
                                    │                                   │
                                    └───────────────────────────────────┘

Figure 1: RemoteAgentLoop Architecture
```

#### 1. **RemoteAgentLoop** (`@register("remote_agent")`)

Extends `AgentLoopBase` to:
- Initialize singleton `LLMProxyServer` on first invocation
- Generate unique `trial_id` for each rollout trajectory
- Submit tasks to remote agent server with proxy URL configuration
- Reconstruct `AgentLoopOutput` from recorded session data

**Key Methods**:
```python
async def run(self, sampling_params: dict[str, Any], **kwargs) -> AgentLoopOutput:
    # 1. Ensure proxy singleton is running
    proxy, proxy_url = await get_or_create_proxy(...)
    
    # 2. Register session with unique trial_id
    trial_id = kwargs["instance_id"] + "-" + shortuuid.uuid()
    proxy.register_session(trial_id)
    
    # 3. Build agent's LLM endpoint URL
    agent_base_url = f"http://{local_ip}:{port}/{trial_id}/v1"
    
    # 4. Submit task to remote harbor server
    result = await client.submit_async(
        task_path=...,
        agent_name="swe-agent",
        environment_overrides={
            "OPENAI_BASE_URL": agent_base_url,
        },
    )
    
    # 5. Collect recorded session data
    session = proxy.get_session_data(trial_id)
    
    # 6. Reconstruct verl-compatible output
    return self._reconstruct_output(session, initial_messages)
```

#### 2. **LLMProxyServer** (Ray Singleton)

FastAPI-based HTTP proxy that:
- Holds vLLM `server_handles` for direct access to rollout servers
- Serves OpenAI-compatible `/v1/chat/completions` endpoints
- Records token_ids, logprobs, and messages per `trial_id`
- Implements sticky-session load balancing across vLLM replicas

**URL Pattern**:
```
POST http://proxy_host:proxy_port/{trial_id}/v1/chat/completions
```

**Session Recording**:
```python
@dataclass
class SessionRecord:
    turns: list[TurnRecord]
    
@dataclass  
class TurnRecord:
    request_messages: list[dict]      # Messages sent to LLM
    completion_token_ids: list[int]   # Generated tokens
    completion_logprobs: list[float]  # Log probabilities
    finish_reason: str
```

#### 3. **AgentRunClient**

HTTP client for submitting agent runs to remote servers:
- Archives and uploads local task directories
- Supports both sync and async submission
- Batch processing for parallel rollouts
- Configurable timeouts for long-running agents

---

## Configuration Example

```yaml
# config.yaml
actor_rollout_ref:
  rollout:
    name: "agent_loop"
    agent:
      default_agent_loop: "remote_agent"
      
    remote_agent:
      # Remote agent server configuration
      harbor_server_url: "http://harbor-server:8080"
      harbor_timeout: 1800.0
      
      # Agent specification
      agent_name: "swe-agent"
      agent_import_path: null  # Or custom agent import path
      model_name: "openai/qwen-plus"
      
      agent_kwargs:
        system_prompt: "You are a helpful coding assistant..."
        
      # Proxy configuration
      proxy_host: "0.0.0.0"
      proxy_port: 0  # Auto-select free port
      
    prompt_length: 4096
    response_length: 8192
    temperature: 0.7
    top_p: 0.95
```

---

### Traffic Flow

1. **Initialization**:
   - `RemoteAgentLoop` worker starts → creates `LLMProxyServer` singleton
   - Proxy binds to auto-selected port, registers with Ray

2. **Rollout Execution**:
   ```
   RemoteAgentLoop.run() 
     ├─► Create trial_id (unique per trajectory)
     ├─► Start LLMProxyServer (if not running)
     ├─► Call remote agent with proxy URL
     │   └─► Agent receives: OPENAI_BASE_URL=http://proxy:port/trial_id/v1
     │
     ├─► Agent makes LLM calls during execution
     │   └─► POST http://proxy:port/trial_id/v1/chat/completions
     │       ├─► Proxy records request messages
     │       ├─► Calls vLLM via server_handle.generate.remote()
     │       ├─► Records completion token_ids + logprobs
     │       └─► Returns response to agent
     │
     └─► After agent completes:
         ├─► Fetch session data from proxy
         ├─► Reconstruct AgentLoopOutput
         └─► Return to trainer for PPO update
   ```

3. **Token Reconstruction**:
   ```python
   def _reconstruct_output(self, session, initial_messages):
       # Initial prompt tokens
       prompt_ids = self._tokenize_messages(initial_messages)
       
       # For each LLM turn:
       for turn in session.turns:
           # LLM completions → mask=1, actual logprobs
           response_ids.extend(turn.completion_token_ids)
           response_mask.extend([1] * len(turn.completion_token_ids))
           response_logprobs.extend(turn.completion_logprobs)
           
           # Inter-turn content (tool/user) → mask=0, logprobs=0.0
           if has_tool_responses:
               tool_ids = self._tokenize_tool_messages(tool_messages)
               response_ids.extend(tool_ids)
               response_mask.extend([0] * len(tool_ids))
               response_logprobs.extend([0.0] * len(tool_ids))
       
       return AgentLoopOutput(...)
   ```

---

### 1. **Zero-Code-Change Integration**
External agents only need to support OpenAI SDK — no VeRL modifications required.

### 2. **Full Trajectory Capture**
Records complete conversation history with token-level metadata for RL training.

### 3. **Infrastructure Isolation**
Agents run in their own environments (Docker, K8s, separate Ray cluster) without contaminating VeRL training setup.

### 4. **Scalable Architecture**
- Single proxy serves multiple concurrent agent loops
- Sticky-session load balancing across vLLM replicas
- Stateless design enables horizontal scaling

### 5. **Debugging & Observability**
All LLM interactions are logged and queryable via session APIs:
```bash
GET /sessions/{trial_id}  # Retrieve complete trajectory
```

---

## Comparison with Existing AgentLoop Types

| Feature | ToolAgentLoop | CollabLLMAgentLoop | **RemoteAgentLoop** |
|---------|---------------|-------------------|---------------------|
| Tool Execution | Local Python tools | Local tools | Remote agent manages tools |
| Agent Location | Same Ray cluster | Same Ray cluster | **Separate Ray/K8s cluster** |
| LLM Connection | Direct vLLM call | Direct vLLM call | **Via HTTP proxy** |
| Token Capture | Native | Native | **Recorded by proxy** |
| Environment | Shared with VeRL | Shared with VeRL | **Isolated (Docker/K8s)** |
| Agent Framework | Custom | Custom | **Any OpenAI-compatible agent** |

---

### Unit Tests
- [x] Proxy session recording (`test_proxy_recorder.py`)
- [x] Token reconstruction logic (`test_output_reconstruction.py`)
- [x] Thread-safe singleton creation (`test_proxy_singleton.py`)

### Integration Tests
- [ ] End-to-end SWE-bench training run
- [ ] Multi-worker concurrent proxy access
- [ ] Fault tolerance (agent crash, network partition)

### Performance Benchmarks
- [ ] Throughput: rollouts/second vs. native ToolAgentLoop
- [ ] Latency breakdown: proxy overhead, network RTT
- [ ] Memory footprint: session buffer growth

---

## Future Enhancements

---

## Related Issues

---

## References

- Implementation: `/sdk/agent_loop/remote_agent_loop.py`
- Proxy Server: `/sdk/proxy/server.py`
- Client SDK: `/sdk/client.py`
- Architecture Diagram: `/docs/images/remote_agent_loop_architecture.png`

---

## Checklist

- [x] Reference implementation provided
- [x] Architecture documentation
- [ ] Unit tests for core components
- [ ] Integration test suite
- [ ] Performance benchmarks
- [ ] User guide for configuring external agents
- [ ] CI/CD pipeline for automated testing

### Current Limitations

VeRL currently provides several AgentLoop implementations:

1. **SingleTurnAgentLoop** - Single-turn chat completion only
2. **ToolAgentLoop** - Multi-turn tool execution with local tools
3. **ReactAgentLoop** - LangGraph-based ReAct pattern
4. **CollabLLMAgentLoop** - Interaction with simulated users via LLM APIs

However, **none of these support**:

- Running agents in external/distributed environments (separate Ray/K8s clusters)
- Capturing token-level data (token_ids, logprobs) from black-box LLM calls
- Integrating production agents that manage their own LLM connections
- RL training of agents that cannot be modified to use VeRL's internal APIs

### Your contribution

I can submit a PR to implement the feature and provide a user guide.

**Comment by Kirrito-k423** [comment]
This is an excellent and well-structured feature request! 🎉

**Summary:**

You're proposing a `RemoteAgentLoop` that enables VeRL to integrate with external agents running in separate Ray or Kubernetes clusters, with a proxy-based architecture for capturing token-level training data.

**Key Benefits:**

1. **Zero-code-change integration**: External agents only need OpenAI SDK support
2. **Infrastructure isolation**: Agents run in their own environments
3. **Full trajectory capture**: Records complete conversation history with token metadata
4. **Scalable architecture**: Single proxy serves multiple concurrent agent loops

**Implementation Evaluation:**

The proposed architecture is sound:
- `LLMProxyServer` as a Ray singleton for traffic interception
- Session-based recording with `trial_id` for trajectory tracking
- Token reconstruction logic that properly masks non-LLM-generated content

**Potential Challenges:**

1. **Latency overhead**: The HTTP proxy layer adds RTT to each LLM call
2. **Memory management**: Session buffers may grow large for long trajectories
3. **Error handling**: Network partitions and agent crashes need robust recovery

**Recommendation:**

Given the comprehensive design and clear use cases, this would be a valuable addition to VeRL. I recommend:

1. Start with a draft PR to `verl/experimental/` directory
2. Include the unit tests you've already implemented
3. Add a simple example using a mock external agent

The community would benefit greatly from this capability, especially for training agents like SWE-agent.

Would you like to proceed with submitting a PR? I can help review the implementation.

**Comment by wuxibin89** [comment]
Yes, we're planing to provide an agent gateway to make every agent framework seamless integrated to verl for RL training.
- OpenAI chat completion API: Conversion between messages and token IDs, tool parser to extract JSON/XML function calls.
- Trajectory tracking: Automatically detecting and switching to a new trajectory for scenarios that message prefix changes, e.g agent skill switch, context compression, and sub-agent, etc.
<img width="500" height="340" alt="Image" src="https://github.com/user-attachments/assets/a85d2f1c-f542-4b2e-8534-e6f474520152" />

**Comment by KunWuLuan** [comment]
> Start with a draft PR to verl/experimental/ directory

Sure. WIP. Hope to complete in next week.

**Comment by KunWuLuan** [comment]
https://github.com/verl-project/verl-recipe/pull/83#pullrequestreview-4096206682

The demo code have been uploaded.
